sdc/extensions/sdc_hashmap_ext.py

```diff
@@ -0,0 +1,1123 @@
+# EXEMPLARY, OR CONSEQUENTIAL DAMAGES (INCLUDING, BUT NOT LIMITED TO,
+# PROCUREMENT OF SUBSTITUTE GOODS OR SERVICES; LOSS OF USE, DATA, OR PROFITS;
+# OR BUSINESS INTERRUPTION) HOWEVER CAUSED AND ON ANY THEORY OF LIABILITY,
+# WHETHER IN CONTRACT, STRICT LIABILITY, OR TORT (INCLUDING NEGLIGENCE OR
+# OTHERWISE) ARISING IN ANY WAY OUT OF THE USE OF THIS SOFTWARE,
+# EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.
+# *****************************************************************************
+
+import llvmlite.binding as ll
+import llvmlite.llvmpy.core as lc
+import numba
+import numpy as np
+import operator
+import sdc
+
+from sdc import hstr_ext
+from glob import glob
+from llvmlite import ir as lir
+from numba import types, cfunc
+from numba.core import cgutils
+from numba.extending import (typeof_impl, type_callable, models, register_model, NativeValue,
+                             lower_builtin, box, unbox, lower_getattr, intrinsic,
+                             overload_method, overload, overload_attribute)
+from numba.cpython.hashing import _Py_hash_t
+from numba.core.imputils import (impl_ret_new_ref, impl_ret_borrowed, iternext_impl, RefType)
+from numba.cpython.listobj import ListInstance
+from numba.core.typing.templates import (infer_global, AbstractTemplate, infer,
+                                         signature, AttributeTemplate, infer_getattr, bound_function)
+from numba import prange
+
+from sdc.str_ext import string_type
+from sdc.str_arr_type import (StringArray, string_array_type, StringArrayType,
+                              StringArrayPayloadType, str_arr_payload_type, StringArrayIterator,
+                              is_str_arr_typ, offset_typ, data_ctypes_type, offset_ctypes_type)
+from sdc.utilities.sdc_typing_utils import check_is_array_of_dtype
+
+from numba.typed.typedobjectutils import _as_bytes
+from sdc import hconc_dict
+from sdc.extensions.sdc_hashmap_type import (ConcurrentDict, ConcurrentDictType,
+                                             ConcDictKeysIterableType, ConcDictIteratorType,
+                                             ConcDictItemsIterableType, ConcDictValuesIterableType)
+from numba.extending import register_jitable
+
+from sdc.extensions.sdc_hashmap_type import SdcTypeRef
+from sdc.utilities.sdc_typing_utils import TypingError, TypeChecker, check_types_comparable
+from itertools import product
+
+from numba.typed.dictobject import _cast
+
+
+def gen_func_suffixes():
+    key_suffixes = ['int32_t', 'int64_t', 'voidptr']
+    val_suffixes = ['int32_t', 'int64_t', 'float', 'double', 'voidptr']
+    return map(lambda x: f'{x[0]}_to_{x[1]}',
+               product(key_suffixes, val_suffixes))
+
+
+def load_native_func(fname, module, skip_check=None):
+    for suffix in gen_func_suffixes():
+        if skip_check and skip_check(suffix):
+            continue
+        full_func_name = f'{fname}_{suffix}'
+        ll.add_symbol(full_func_name,

```

Comment:
Is there a reason, we are using `ll.add_symbol` and not ctypes binding, so we can directly call functions from jit code?

Like it is done here:
https://github.com/IntelPython/sdc/blob/master/sdc/functions/sort.py#L55

Reply:
No, there's no particular reason for this (except it was easier to write LLVM codegen that way - the function call inserted there directly refers to the function name being loaded, so there's 1:1 mapping and it's easier to navigate the code). I think we may refactor this and re-use common ctypes/bind approach later.